### PR TITLE
Mockito #1013: Defines and implements API for static mocking.

### LINKED ISSUE #1013: Enable mocking static methods in Mockito
# Static methods mocking with Mockito

This is a placeholder ticket for enabling mocking static methods in Mockito. Example action items that are totally negotiable and can ran in parallel. We're looking for someone who can lead this effort.

- [x] Research + discuss whether it is a good idea to enable static methods mocking in Mockito. The theory is that it is useful for legacy code, which is most code in the world.
- [x] Research on how other mocking frameworks do that and whether it is considered useful feature for their users.
- [x] Design and present for discussion an API for static mocking (slightly relevant ticket: #643)
- [x] Work with @raphw / ByteBuddy to come up with hacky prototype (the hackier, the better!)
- [x] Mold the prototype with the API, remove enough rough edges so that the feature is good enough for incubating rollout
- [ ] SHIP IT!

### PR DESCRIPTION
**Pull request by raphw**
Fixes #1013 - allows for static method mocking in Mockito.

### CODE REVIEW

**Comment by raphw** [comment]
I thought about it but I am pretty confident that this is what it's gonna be! I suggest we put the APIs as incubating as we have done with the inline mock maker. If we commit to the API, even in a separate artifact, I think the Android folks will still go for it and have a similar expectation on it's stability.

You think a separate artifact would be cleaner?

**Comment by TimvdLippe** [comment]
I mostly prefer a conservative route if we have that option. But keeping as Incubating is okay with me.

**Comment by raphw** [comment]
I think I'd prefer the single artifact. It's fully opt-in and does not touch any existing code. I think this way it will be used more and we'll find out the adoption quicker.

**Comment by mockitoguy** [comment]
I just saw this and cannot believe my eyes :-) Will try to review shortly!!!

**Comment by raphw** [comment]
Awesome, just discovered a remaining recursion on `thenCallRealMethod` but I think I already know how to fix it.

**Comment by raphw** [comment]
And it works, too!

File: src/main/java/org/mockito/MockedStatic.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        }
+    }
+
+    private void assertNotClosed() {
+        if (closed) {
+            throw new MockitoException(
+                    join(
+                            "The static mock created at",
+                            location.toString(),
+                            "is already resolved and cannot longer be used"));
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/MockedStatic.java:200]:
```suggestion
                            "is already resolved and can no longer be used"));
```

File: src/main/java/org/mockito/Mockito.java [hunk 2]
```diff
@@ -1541,6 +1542,29 @@
...
+ *
+ * To make sure a static mock remains temporary, it is recommended to define the scope within a try-with-resources construct.
+ * In the following example, the <code>Foo</code> type's static method would return <code>foo</code> unless mocked:
+ *
+ * <pre class="code"><code class="java">
+ * assertEquals("foo", Foo.method());
+ * try (MockedStatic<Foo> mocked = mockStatic(Foo.class)) {
+ * mocked.when(Foo::method).thenReturn("bar");
+ * assertEquals("bar", Foo.method());
+ * mocked.verify(Foo::method);
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/Mockito.java:1560]:
```suggestion
 *     mocked.when(Foo::method).thenReturn("bar");
 *     assertEquals("bar", Foo.method());
 *     mocked.verify(Foo::method);
```

File: src/main/java/org/mockito/MockitoAnnotations.java [hunk 3]
```diff
@@ -37,8 +37,14 @@
  *
  *   public class SampleBaseTestCase {
  *
+ *       private AutoCloseable closeable;
+ *
  *       &#064;Before public void initMocks() {
- *           MockitoAnnotations.initMocks(this);
+ *           closeable = MockitoAnnotations.initMocks(this);
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/MockitoAnnotations.java:43]: This is potentially a breaking change, if compilations downstream require the result of any `AutoCloseable` to be used. I would have to check for example internally if that is the case. Can we maybe separate this out into a separate method called `MockitoAnnotations.initWithStaticMocks`? That would make it clearer that a particular test is allowed to use static mocks. By using a separate method, users can write static analyses that verify that `.close` is called.

File: src/main/java/org/mockito/internal/creation/bytebuddy/InlineByteBuddyMockMaker.java [hunk 4]
```diff
@@ -300,19 +314,41 @@ public MockHandler getHandler(Object mock) {
     public void resetMock(Object mock, MockHandler newHandler, MockCreationSettings settings) {
         MockMethodInterceptor mockMethodInterceptor =
                 new MockMethodInterceptor(newHandler, settings);
-        mocks.put(mock, mockMethodInterceptor);
-        if (mock instanceof MockAccess) {
-            ((MockAccess) mock).setMockitoInterceptor(mockMethodInterceptor);
+        if (mock instanceof Class<?>) {
+            Map<Class<?>, MockMethodInterceptor> interceptors = mockedStatics.get();
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/creation/bytebuddy/InlineByteBuddyMockMaker.java:318]: I am not really following this refactoring. Could you elaborate on why this is necessary? Would it maybe be possible to separate these changes into its own PR to ease reviewing?

File: src/main/java/org/mockito/MockitoAnnotations.java (hunk 3 above)
Comment (raphw) [reply src/main/java/org/mockito/MockitoAnnotations.java:43]: I thought about this, too. It is source compatible but not binary compatible. However, I do not think that this is a very commonly used API by derivative tools and direct users do not have to apply any changes until using static mocks in their custom tooling what likely requires them to revisit this API. In a way, explicit life-cycles are a new concept in Mockito that get introduced by this change, it will require some upstream adoption and in such a case, I think that breakage is the most explicit way to communicate this to users; I always prefer this over silent or runtime failure. I suggest therefore to make this visible in the release notes and by the versioning scheme via releasing a new feature version when merging this.

File: src/main/java/org/mockito/internal/creation/bytebuddy/InlineByteBuddyMockMaker.java (hunk 4 above)
Comment (raphw) [reply src/main/java/org/mockito/internal/creation/bytebuddy/InlineByteBuddyMockMaker.java:318]: It's easier when looking at the raw file, the changefile cuts parts of the interesting segment. Basically, to avoid breaking the existing implementation, static mocks are represented by the `Class` value for the static mock. This is also done in the listerners etc. what is possible since it is currently impossible to mock a `Class` instance, even with the inline-mock maker.

We cannot store static mocks in the same concurrent map as the regular mocks since they are thread local. Therefore, we always need to make a distinction if a mock is an instance of `Class` or not to see if a static or a regular mock is requested. The advantage of this solution is that the current API can mainly remain untouched, keeping the changeset to a minimum.

**Comment by raphw** [comment]
I thought about splitting the `MockitoAnnotations` into offering two methods. However, since we introduce a life-cycle, the "old" method can no longer be used safely. It's a rather uncommon API to use manually, therefore I believe that a breaking change is a good way to notify users that they need to pay attention. Most people do however use the JUnit runner, rule and extension, I think and won't be affected.

I considered splitting up the PR but that way I could not implement static mocks as a running feature without putting all pieces together. I therefore suggest that we release it as one change, that way Mockito stays consistent in its offered features and we hopefully get frictionless adoption. I am afraid that this change has a certain inherent complexity that would not go away even when splitting up.

**Comment by TimvdLippe** [comment]
> It's a rather uncommon API to use manually, therefore I believe that a breaking change is a good way to notify users that they need to pay attention.

I have seen numerous usages in Google. Before we make this change, I would like to verify that this is indeed safe to do. I am not sure when I have time to do so, but I think someday next week should be possible.

**Comment by raphw** [comment]
I have not considered wide usage. In this case, I think it's better to deprecate `initMocks` in favor of `openMocks` where the former does not misbehave unless you are using the new static mocks. This way, there is no API breakage and Mockito continues to well-behave for unchanged user code.

I think such a life-cycle will be useful for other cases, too. For instance, I think it opens a solution for https://github.com/mockito/mockito/issues/1802 where I otherwise cannot see a solution.

**Comment by TimvdLippe** [comment]
`initMocks` is used widely in the Android ecosystem, as it (allegedly) has better integration with ClassLoader like Robolectric. I think since then, we have addressed the issues in Mockito, but the usage of `initMocks` remains prevalent.

File: src/main/java/org/mockito/Captor.java [hunk 5]
```diff
@@ -15,9 +15,16 @@
  *
  *    &#64;Captor ArgumentCaptor&lt;AsyncCallback&lt;Foo&gt;&gt; captor;
  *
+ *    private AutoCloseable closeable;
+ *
  *    &#64;Before
- *    public void init(){
- *       MockitoAnnotations.initMocks(this);
+ *    public void open() {
+ *       MockitoAnnotations.openMocks(this);
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/Captor.java:22]:
```suggestion
 *       closeable = MockitoAnnotations.openMocks(this);
```

Comment (raphw) [reply src/main/java/org/mockito/Captor.java:22]: Awesome! We could also update Byte Buddy some time soon to get initial Java 16 support.

File: src/main/java/org/mockito/plugins/MockMaker.java [hunk 6]
```diff
@@ -108,6 +109,26 @@
...
+     * @param settings Mock creation settings like type to mock, extra interfaces and so on.
+     * @param handler See {@link org.mockito.invocation.MockHandler}.
+     *                <b>Do not</b> provide your own implementation at this time. Make sure your implementation of
+     *                {@link #getHandler(Object)} will return this instance.
+     * @param <T> Type of the mock to return, actually the <code>settings.getTypeToMock</code>.
+     * @return A control for the static mock.
+     * @since 3.4.0
+     */
+    @Incubating
+    <T> StaticMockControl<T> createStaticMock(
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/plugins/MockMaker.java:129]: This is a breaking change and breaks Dexmaker. I recommend adding a default implementation with ```throw new MockitoException("The subclass byte code generator cannot create static mocks");```

File: src/main/java/org/mockito/MockedStatic.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    void close();
+
+    /**
+     * Releases this static mock and is non-operational if already released.
+     */
+    void closeOnDemand();
+
+    @FunctionalInterface
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/MockedStatic.java:74]: This is not available on Android builds and therefore will fail. I think we can remove this annotation and the rest will still work?

Comment (raphw) [reply src/main/java/org/mockito/MockedStatic.java]: You are right, I commented it out. It is mainly for documentation purposes and to fail compilation if the interfaces does not define a single abstract method.

File: src/main/java/org/mockito/plugins/MockMaker.java (hunk 6 above)
Comment (raphw) [reply src/main/java/org/mockito/plugins/MockMaker.java]: You are right, I added a default method which throws an explaining exception. I also had to move the static mock control into a non-internal package. Otherwise nobody else could implement the API.

File: src/main/java/org/mockito/Captor.java (hunk 5 above)
Comment (TimvdLippe) [reply src/main/java/org/mockito/Captor.java:22]: Could you apply this suggestion?

File: src/main/java/org/mockito/MockedStatic.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    void close();
+
+    /**
+     * Releases this static mock and is non-operational if already released.
+     */
+    void closeOnDemand();
+
+    // @FunctionalInterface - not supported on Android
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/MockedStatic.java:74]:
```suggestion
```

**Comment by TimvdLippe** [comment]
@raphw You also need to run `./gradlew spotlessApply` locally to format the sourcecode.

File: src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java [hunk 9]
```diff
@@ -41,6 +50,29 @@ public static Object processAnnotationForMock(Mock annotation, Class<?> type, St
...
+            return Mockito.mockStatic(inferStaticMock(genericType.get(), name), mockSettings);
+        } else {
+            return Mockito.mock(type, mockSettings);
+        }
+    }
+
+    private static Class<?> inferStaticMock(Type type, String name) {
+        if (type instanceof ParameterizedType) {
+            ParameterizedType parameterizedType = (ParameterizedType) type;
+            return (Class<?>) parameterizedType.getRawType();
```
Comment (andrei-ivanov) [inline src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: Shouldn't this use `getActualTypeArguments()[0]` instead?
It's failing for me now since `getRawType()` returns `MockedStatic`

Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: Do you mind opening a PR with regression test + fix? Thanks!

Comment (andrei-ivanov) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: Hmm, seems this was easy to fix locally with my suggestion, but I now get another error, which I don't know why it happens:
```
org.mockito.exceptions.misusing.NotAMockException: Argument passed to Mockito.mockingDetails() should be a mock, but is an instance of class java.lang.Class!

	at org.mockito.junit.jupiter.MockitoExtension.afterEach(MockitoExtension.java:186)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAfterEachCallbacks$11(TestMethodTestDescriptor.java:255)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$316/0000000000000000.invoke(Unknown Source)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$12(TestMethodTestDescriptor.java:271)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$1809/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$13(TestMethodTestDescriptor.java:271)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$314/0000000000000000.accept(Unknown Source)
```
The class with the static method ends up there...

Comment (raphw) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: I isolated the error. The problem is that the mocks for static methods are scoped. This means that they are released at the end of the try-with-resources block what implies that the mocks are no longer valid at the end of the method resulting in this error. 

@TimvdLippe For now I think the easiest is to exclude static mocks from the post processing. The more correct approach would be to execute the "after each" logic upon releasing the mock rather then at the end of the test. I will look into it but need to understand the post-processing first.

Comment (andrei-ivanov) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: Btw, my setup was with and without a try-with-resources block, since I understood from the documentation of `@Mock` that it should work with that as well, which would simplify things (and since I already had other fields that were annotated with `@Mock`)

Comment (andrei-ivanov) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]:
```java
@ExtendWith(MockitoExtension.class)
public class TenantScopeTest {
    @Mock
    private UserPrincipal principal;

    @Mock
    private MockedStatic<UserPrincipalUtil> userPrincipalUtil;

    @Test
    public void opsShouldBeCachedInstancesByNameAndTenant() {
        String tenant = "x";
        when(principal.getMainTenant()).thenReturn(tenant);

        userPrincipalUtil.when(UserPrincipalUtil::requireUserPrincipal).thenReturn(principal);

        //....
}
```

Comment (raphw) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: @andrei-ivanov I have not had the time to test it yet (we should have unit tests for the runners using both mock makers) but maybe the 'runner-fix' branch already fixes the problem.

Comment (andrei-ivanov) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: I still get the `NotAMockException` with the `runner-fix` branch

Comment (andrei-ivanov) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: Let me know when you want to do run some more tests, I'm eager to get rid of PowerMock 😁

Comment (jacky1193610322) [reply src/main/java/org/mockito/internal/configuration/MockAnnotationProcessor.java:64]: I also have this problem, when the mockMaker is InlineByteBuddyMockMaker, I use SpyBean to spy a spring bean, and then verify the bean, it throws NotAMockException. I debug and find the class type is not contains mockito, but it contains when I use the ByteBuddyMockMaker.

**Comment by jacky1193610322** [comment]
I also have this problem, when the mockMaker is InlineByteBuddyMockMaker, I use SpyBean to spy a spring bean, and then verify the bean, it throws NotAMockException. I debug and find the class type is not contains mockito, but it contains when I use the ByteBuddyMockMaker. and it seems not a spring bug.
